### ISSUE TITLE
Bug: [no-unnecessary-condition] incorrect behaviour in if statement when checking object value by key

### ISSUE THREAD

**Issue by pauldasslier** [typescript-eslint/typescript-eslint#5236]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=JYOwLgpgTgZghgYwgAgJIAkD2BbCB3TKAa2QG8AoZZYAEwC5kQBXbAI2nIF9zyFMQAzmGQALHPkJEGGcQWIBtALrIAvMiUBuHn0HCA5lExMADhBpZccqWXW0GQqKD2LpFiQuWdVZTlvJjLSQA6GEIAUUQRAAookTwASlUAPjJKahhkKIBCAyNTc1lJeTig2kVEiioqXJMzNytivFKaZTVGxS0qLwgAGwEUSqqa-PqikrKg4yYBaLj4zuRuTnmgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iQANODABfEDqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
interface IHomework {
  id: number
}

const homework: IHomework[] = [];

const groupedHomework: { [id: string]: IHomework[] } = {};

homework.forEach((hw) => {
  if (!groupedHomework[hw.id]) {
    groupedHomework[hw.id] = [hw];
  } else {
    groupedHomework[hw.id].push(hw);
  }
});
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error",
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Without error

### Actual Result

Has an error

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.29.0` |
| `@typescript-eslint/parser` | `5.29.0` |
| `TypeScript` | `4.7.2` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by bradzacher** [comment]
Same as #2941 and other issues.
